### DISCUSSION TITLE
Moving Tables and Figures Produced by Executable Code Blocks in Quarto

### DISCUSSION THREAD

**Discussion by TomBener** [quarto-dev/quarto-cli#9934] [Q&A] [answered]
Hello Quarto community,

I have been working on a Lua filter designed to move tables and figures to the end of a DOCX document, as required by some academic journal formats. The filter functions well for general tables and external figures processed by Pandoc. Below is the Lua filter I am currently using:

```lua
--- move-floats.lua
--- Move tables and figures to the end of docx

-- Function to create a page break
function create_page_break()
    return pandoc.RawBlock('openxml', '<w:p><w:r><w:br w:type="page"/></w:r></w:p>')
end

-- Function to create a placeholder for the original position
function create_placeholder(element_type, index)
    local text = element_type == 'Table' and pandoc.Strong({ pandoc.Str("TABLE " .. index .. " HERE") }) or
        pandoc.Strong({ pandoc.Str("FIGURE " .. index .. " HERE") })
    return pandoc.Para { text }
end

-- Function to move elements to the end of the document
function move_floats_to_end(doc, element_type, header_title)
    local blocks = {}
    local elements = {}

    -- Separate the specified elements from other blocks
    local index = 1
    for _, el in ipairs(doc.blocks) do
        if el.t == element_type then
            -- Insert a placeholder in the original position
            table.insert(blocks, create_placeholder(element_type, index))
            table.insert(elements, el)
            index = index + 1
        else
            table.insert(blocks, el)
        end
    end

    -- Add a heading for the elements if any exist
    if #elements > 0 then
        -- Insert page break before heading
        table.insert(blocks, create_page_break())
        -- Create an unnumbered header
        local header = pandoc.Header(1, pandoc.Str(header_title))
        header.classes:insert("unnumbered")
        table.insert(blocks, header)
        -- Add elements and a page break after each one, except for the last element
        for i, el in ipairs(elements) do
            table.insert(blocks, el)
            if i < #elements then
                table.insert(blocks, create_page_break())
            end
        end
    end

    return blocks
end

-- Main function to handle the document
function Pandoc(doc)
    -- Move tables to the end
    local blocks_with_tables_moved = move_floats_to_end(doc, 'Table', 'Tables')
    -- Create a new document to include the moved tables
    local doc_with_tables_moved = pandoc.Pandoc(blocks_with_tables_moved, doc.meta)
    -- Move figures to the end
    local final_blocks = move_floats_to_end(doc_with_tables_moved, 'Figure', 'Figures')
    -- Insert a page break after the last block
    table.insert(final_blocks, create_page_break())
    -- Return the modified document
    return pandoc.Pandoc(final_blocks, doc.meta)
end
```

**Problem:**

While the filter works effectively for general tables and figures, it encounters issues when dealing with tables and figures produced by executable code blocks within Quarto documents. Specifically:

1. **Failure to Move Executable Block Outputs:** The Lua filter does not relocate tables and figures generated by code blocks to the end of the document. They remain in their original positions or some mess, leading to a mix of placed and unplaced elements.

2. **Cross-Reference Corruption:** The relocation process seems to corrupt cross-references. References to these tables and figures break, resulting in issues with the integrity and readability of the document.

**Steps to Reproduce:**

1. Create a Quarto document with executable code blocks that generate tables and figures, for example:

````md
---
title: Moving Tables and Figures Produced by Executable Code Blocks in Quarto
execute:
  cache: true
  freeze: auto
  echo: false
  warning: false
format:
  docx:
    filters:
      - move-floats
---

Texts with executable code blocks in Quarto file `index.qmd`.

| Col1 | Col2 | Col3 |
|------|------|------|
| A    | B    | C    |
| E    | F    | G    |
| A    | G    | G    |

: My Caption {#tbl-letters}

See @tbl-letters.

```{r}
#| message: false
#| fig-cap: "A demo figure"
#| label: carfig

library(ggplot2)
mtcars2 <- mtcars
mtcars2$am <- factor(
  mtcars$am, labels = c('automatic', 'manual')
)
ggplot(mtcars2, aes(hp, mpg, color = am)) +
  geom_point() + geom_smooth() +
  theme(legend.position = 'bottom')
```

```{r}
#| label: tbl-example
#| tbl-cap: "Example"
#| tbl-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout-ncol: 2

library(knitr)
kable(head(cars))
kable(head(pressure))
```

```{r}
#| label: fig-charts
#| fig-cap: "Charts"
#| fig-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout-ncol: 2

plot(cars)
plot(pressure)
```
````

2. Apply the Lua filter to move the tables and figures to the end of the document:

```shell
quarto render index.qmd --to docx
```

3. Observe that the tables and figures from code blocks do not move and that cross-references are broken.

**Questions:**

1. How can the Lua filter be adjusted or supplemented to successfully move tables and figures produced by executable code blocks?
2. Is there a recommended approach in Quarto for maintaining cross-references integrity after moving such elements?
3. Are there any known limitations or workarounds for this issue in Quarto or Pandoc that might help resolve these problems?

Any guidance or suggestions on how to address these issues would be greatly appreciated. Thank you for your help!

**Comment by mcanouil** [comment]
Could you add the format and the filters YAML keys in your example? See <https://quarto.org/docs/extensions/filters.html>.
Also, you might want to ensure where/when your filter applies, see <https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases>.

**Reply by TomBener** [reply]
Thanks for your reminding. I have updated the description.

**Reply by mcanouil** [reply]
Your example does not help to show what your filter is supposed to do as nearly all figures/tables are already at the end.

Did you try using the native format to look at the AST and debug your filter? https://quarto.org/docs/extensions/lua.html#native-format

> The filter functions well for general tables and external figures 

Well it does not work at all at least in 1.5.
I made the example a bit more easier to see if the table is moved or not.

<table>
<tr><th>Quarto document</th><th>HTML</th></tr>
<tr><td>

````qmd
---
title: Moving Tables and Figures Produced by Executable Code Blocks in Quarto
format:
  docx:
    filters:
      - move-floats.lua
---

Texts with executable code blocks in Quarto file `index.qmd`.

| Col1 | Col2 | Col3 |
|------|------|------|
| A    | B    | C    |
| E    | F    | G    |
| A    | G    | G    |

: My Caption {#tbl-letters}

See @tbl-letters.

## Something

{{< lipsum 1 >}}

````

</td><td>

<img width="425" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/e72da32f-3b0f-4c0c-8667-4a7e971b415a">

</td></tr>
</table>

**Reply by TomBener** [reply]
Thanks. Indeed, the Lua filter above does not work in the example. I tried to update it:

```lua
--- Move tables and figures to the end of docx

-- Function to create a page break
function create_page_break()
    return pandoc.RawBlock('openxml', '<w:p><w:r><w:br w:type="page"/></w:r></w:p>')
end

-- Function to create a placeholder for the original position
function create_placeholder(element_type, index)
    local text = element_type == 'Table' and pandoc.Strong({ pandoc.Str("TABLE " .. index .. " HERE") }) or
        pandoc.Strong({ pandoc.Str("FIGURE " .. index .. " HERE") })
    return pandoc.Para { text }
end

-- Function to move elements to the end of the document
function move_floats_to_end(doc, element_type, header_title)
    local blocks = {}
    local elements = {}

    -- Function to recursively check for elements in a block
    local function check_block(block)
        if block.t == element_type then
            -- Insert a placeholder in the original position
            table.insert(blocks, create_placeholder(element_type, #elements + 1))
            table.insert(elements, block)
        elseif block.t == 'Div' then
            -- Check each block inside the Div
            for _, inner_block in ipairs(block.content) do
                check_block(inner_block)
            end
        else
            table.insert(blocks, block)
        end
    end

    -- Separate the specified elements from other blocks
    for _, block in ipairs(doc.blocks) do
        check_block(block)
    end

    -- Add a heading for the elements if any exist
    if #elements > 0 then
        -- Insert page break before heading
        table.insert(blocks, create_page_break())
        -- Create an unnumbered header
        local header = pandoc.Header(1, pandoc.Str(header_title))
        header.classes:insert("unnumbered")
        table.insert(blocks, header)
        -- Add elements and a page break after each one, except for the last element
        for i, el in ipairs(elements) do
            table.insert(blocks, el)
            if i < #elements then
                table.insert(blocks, create_page_break())
            end
        end
    end

    return blocks
end

-- Main function to handle the document
function Pandoc(doc)
    -- Move tables to the end
    local blocks_with_tables_moved = move_floats_to_end(doc, 'Table', 'Tables')
    -- Create a new document to include the moved tables
    local doc_with_tables_moved = pandoc.Pandoc(blocks_with_tables_moved, doc.meta)
    -- Move figures to the end
    local final_blocks = move_floats_to_end(doc_with_tables_moved, 'Figure', 'Figures')
    -- Insert a page break after the last block
    table.insert(final_blocks, create_page_break())
    -- Return the modified document
    return pandoc.Pandoc(final_blocks, doc.meta)
end
```

Now it should move some floats, but it made subfigures and subtables messy and captions with labels were not moved. Also, Quarto reported `WARNING (/Applications/quarto/share/filters/main.lua:13808) Unable to resolve crossref @tbl-letters`.

**Reply by mcanouil** [reply]
Are you using the native format to see what you need to move? Or just logging to see the content before quarto process it.
Did you try to execute your filter before "quarto" as documented in the links I shared?

One thing I think could be the issue is all the Quarto processing for cross-reference in which I am not very familiar with.
Maybe @cscheid has some insights to share here on what you try to do.

**Comment by iagopinal** [comment, marked as answer]
Take a look at [figure-table-formatter](https://github.com/iagopinal/figure-table-formatter). It moves the tables, figures and supplements to wherever you indicate. You can cross-reference the figures and tables in the order that they are mentioned in the text, and sort them accordingly. It automatically outputs a file with the sorted tags of the figures and tables. You can use this file to automatically sort the figures in other files using the same filter (handy for supplements).

It would be nice to have the same functionality in the native Quarto, but at least this is a viable alternative now.
